pkg/webhook/admission.go

```diff
@@ -0,0 +1,230 @@
+}
+
+func admissionRequired(ignoredList []string, admissionAnnotationKey string, metadata *metav1.ObjectMeta) bool {
+	// skip special kubernetes system namespaces
+	for _, namespace := range ignoredList {
+		if metadata.Namespace == namespace {
+			glog.Infof("Skip validation for %v for it's in special namespace:%v", metadata.Name, metadata.Namespace)
+			return false
+		}
+	}
+
+	annotations := metadata.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+
+	var required bool
+	switch strings.ToLower(annotations[admissionAnnotationKey]) {
+	default:
+		required = true
+	case "n", "no", "false", "off":
+		required = false
+	}
+	return required
+}
+
+func validationRequired(ignoredList []string, metadata *metav1.ObjectMeta) bool {
+	required := admissionRequired(ignoredList, admissionWebhookAnnotationValidateKey, metadata)
+	glog.Infof("Validation policy for %v/%v: required:%v", metadata.Namespace, metadata.Name, required)
+	return required
+}
+
+// validate deployments and services
+func (whsvr *WebhookServer) validate(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	req := ar.Request
+	status := &v1beta1.AdmissionResponse{}
+
+	if ar.Request.Operation != v1beta1.Delete && req.Kind.Kind != "PersistentVolumeClaim" {
+		status.Allowed = true
+		return status
+	}
+	glog.Infof("AdmissionReview for Kind=%v, Namespace=%v Name=%v UID=%v patchOperation=%v UserInfo=%v",
+		req.Kind, req.Namespace, req.Name, req.UID, req.Operation, req.UserInfo)
+
+	// TODO use raw object once validation webhooks support DELETE request
+	// object as non nil value https://github.com/kubernetes/kubernetes/issues/66536
+	//var pvc corev1.PersistentVolumeClaim
+	//err := json.Unmarshal(req.Object.Raw, &pvc)
+	//if err != nil {
+	//	glog.Errorf("Could not unmarshal raw object: %v, %v", err, req.Object.Raw)
+	//	status.Allowed = false
+	//	status.Result = &metav1.Status{
+	//		Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+	//		Message: err.Error(),
+	//	}
+	//	return status
+	//}
+
+	// fetch the pvc specifications
+	pvc, err := whsvr.KubeClient.CoreV1().PersistentVolumeClaims(req.Namespace).Get(req.Name, metav1.GetOptions{})
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	if !validationRequired(ignoredNamespaces, &pvc.ObjectMeta) {
+		glog.Infof("Skipping validation for %s/%s due to policy check", pvc.Namespace, pvc.Name)
+		return &v1beta1.AdmissionResponse{
+			Allowed: true,
+		}
+	}
+
+	status.Allowed = true
+	cStorVolumes, _ := whsvr.Clientset.OpenebsV1alpha1().CStorVolumes("openebs").List(metav1.ListOptions{})
+	for _, cstorvolume := range cStorVolumes.Items {
+		if cstorvolume.Annotations["openebs.io/source-volume"] == pvc.Spec.VolumeName {
+			status.Allowed = false
```

Comment:
can we do 'break' from here? or list out all the clones on this volume in the error message..

Reply:
@vishnuitta the changes has been already updated ... those are the older commits . thanks :slightly_smiling_face: